### PR TITLE
client: Allow SSL use when communicating with Zookeeper, fixes #382

### LINKED ISSUE #382: Add support for SSL
As SSL is [supported by Zookeeper starting from 3.5.1](https://issues.apache.org/jira/browse/ZOOKEEPER-2125) it would be nice to have it in kazoo.

### PR DESCRIPTION
**Pull request by emonty**
Zookeeper 3.5 supports SSL for client communications, this commit
adds support for it on the Kazoo side.

Note that you need to give the client the key, certificate and CA
files.

Co-Authored-By: Monty Taylor <mordred@inaugust.com>

### CODE REVIEW

**Comment by jeffwidman** [comment]
Why the new PR when you already had #493? Are you aware that you can force push to your PR branch and GitHub will update your PR?

File: kazoo/handlers/utils.py [hunk 1]
```diff
@@ -194,17 +197,41 @@ def create_tcp_connection(module, address, timeout=None):
...
+                                           socket.SOCK_STREAM)
+                conn = context.wrap_socket(module.socket(addrs[0][0]))
+                conn.connect(address)
+                sock = conn
+                break
+            except ssl.SSLError:
+                raise
+        else:
+            try:
+                sock = module.create_connection(address, timeout)
```
Comment (jeffwidman) [inline kazoo/handlers/utils.py]: I noticed this wasn't previously present... was it overlooked prior or is there some reason it's needed now?

**Comment by jeffwidman** [comment]
nudge @emonty ^^

**Comment by Shrews** [comment]
@jeffwidman Pretty sure @emonty opened this one and closed the other so that he could grant permission to other folks to iterate on changes. In the other PR, only the original author (not us) could do that.

File: kazoo/handlers/utils.py (hunk 1 above)
Comment (Shrews) [reply kazoo/handlers/utils.py]: Can you elaborate? Not quite sure what you are referring to. Looks the same as the original author's code, AFAICT.

Comment (jeffwidman) [reply kazoo/handlers/utils.py]: The original code is already copied below in https://github.com/python-zk/kazoo/pull/494/files#diff-6aabced2bdeee5ae3b908caaa504d88eR225

https://github.com/python-zk/kazoo/pull/494/files#diff-6aabced2bdeee5ae3b908caaa504d88eR223 appears to be a copy/paste error.

**Comment by jeffwidman** [comment]
@Shrews hmm, that's an interesting idea and I hadn't noticed they were different. However, no need for another PR, it's possible to flip that on/off using a checkbox on the original PR.

Thank you for reviewing the PR as well, always appreciate another set of eyes.

File: kazoo/handlers/utils.py (hunk 1 above)
Comment (jeffwidman) [reply kazoo/handlers/utils.py]: @emonty  / @Ormod can you fix the above?

File: kazoo/handlers/utils.py [hunk 2]
```diff
@@ -194,17 +197,41 @@ def create_tcp_connection(module, address, timeout=None):
...
+            # Disallow use of SSLv2 and V3 (meaning we require TLSv1.0+)
+            context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+            context.options |= ssl.OP_NO_SSLv2
+            context.options |= ssl.OP_NO_SSLv3
+            context.verify_mode = ssl.CERT_OPTIONAL
+            if ca:
+                context.load_verify_locations(ca)
+            if certfile and keyfile:
+                context.verify_mode = ssl.CERT_REQUIRED
+                context.load_cert_chain(certfile=certfile, keyfile=keyfile)
```
Comment (StephenSorriaux) [inline kazoo/handlers/utils.py]: Would it be possible to also add the keyfile's password?

Comment (StephenSorriaux) [reply kazoo/handlers/utils.py]: Thanks @jeffwidman, I will surely spend time to implement this

Comment (StephenSorriaux) [reply kazoo/handlers/utils.py]: @jeffwidman I just submitted a PR for SSL support https://github.com/python-zk/kazoo/pull/513

**Comment by jeffwidman** [comment]
Closing in favor of #513

**Pull request closed without merging**